pypeit/core/trace.py

```diff
@@ -0,0 +1,1396 @@
+
+    if nleft == 0 and nright == 0 and not flux_valid:
+        # No traces and fluxes are invalid.
+        # TODO: This used to just be a warning, but I'm having it stop
+        # the code if no traces are found and the flux is low.
+        msgs.error('Unable to trace any edges!  Image flux is low; check trace image is correct.')
+
+    # Use the mask to determine the first and last valid pixel column
+    sum_bpm = np.zeros(edge_img.shape[1]) if bpm is None else np.sum(bpm, axis=0) 
+
+    if nleft == 0:
+        # Add a left edge trace at the first valid column
+        msgs.warn('No left edge found. Adding one at the detector edge.')
+        gdi0 = np.min(np.where(sum_bpm[buffer:] == 0)[0]) + buffer
+        _edge_img[:,gdi0] = -1
+
+    if nright == 0:
+        # Add a right edge trace at the last valid column
+        msgs.warn('No right edge found. Adding one at the detector edge.')
+        gdi1 = np.max(np.where(sum_bpm[:-buffer] == 0)[0])
+        _edge_img[:,gdi1] = 1
+
+    return _edge_img
+
+
+# TODO: This needs to be better tested
+def handle_orphan_edges(edge_img, sobel_sig, bpm=None, flux_valid=True, buffer=0, copy=False):
+    """
+    In the case of single left/right traces and multiple matching
+    traces, pick the most significant matching trace and remove the
+    others.
+
+    If *no* left and/or right edge is present, this will add one using
+    :func:`atleast_one_edge`.
+
+    Args:
+        edge_img (`numpy.ndarray`_):
+            Image with edge trace pixels numbered by their associated
+            trace.  Pixels with positive numbers follow right slit edges
+            and negative numbers follow left slit edges.
+        sobel_sig (`numpy.ndarray`_):
+            Image with the significance of the edge detection.  See
+            :func:`detect_slit_edges`.
+        bpm (`numpy.ndarray`_, optional):
+            Integer (0 unmasked; 1 masked) or boolean array indicating
+            bad pixels in the image.  If None, all pixels are considered
+            good.
+        flux_valid (:obj:`bool`, optional):
+            The flux in the image used to construct the edge traces is
+            valid meaning that any problems should not be an issue with
+            the trace image itself.
+        buffer (:obj:`int`, optional):
+            If adding an edge, this is the minimum number of pixels
+            near the detector edge at which to place the edge; see
+            :func:`atleast_one_edge`.
+        copy (:obj:`bool`, optional):
+            Copy `edge_img` to a new array before making any
+            modifications. Otherwise, `edge_img` is modified
+            in-place.
+
+    Returns:
+        `numpy.ndarray`_: The modified trace image, which is either a
+        new array or points to the in-place modification of
+        `edge_img` according to the value of `copy`.
+    """
+    # Get the number of traces
+    nleft, nright = count_edge_traces(edge_img)
+
+#    if nleft == 0 or nright == 0:
```

Comment:
`at_least_one_edge` and `handle_orphan_edges` are nearly the same as what was `core.trace_slits.edgearr_add_left_right` and `core.trace_slits.edgearr_final_left_right`.  When I was testing the code I ran into some problems with how this was behaving between different instruments.

I think we can safely get rid of `atleast_one_edge` because this is basically handled in `EdgeTraceSet.sync`.

In `EdgeTraceSet.initial_trace`, I'm also now skipping the call to `handle_orphan_edges` because of a problem I ran into with DEIMOS long slit data.  The problem was that there's a (bad) hot column that has a more significant edge than the actual edge of the long slit, and the nominal tracing only finds a single opposite edge.  So that means that `handle_orphan_edges` keeps the edge caused by the bad column and we lose the actual edge in the rest of the code.  But if I skip `handle_orphan_edges`, this all gets handled by `EdgeTraceSet.sync`.  I've kept the code around though because I think we need to talk through this, and I think the result is that we now may end up with too many slits.

Long story short, I agree that we may want to remove this eventually, but I think we keep it around for now until we understand if we need to bring this functionality back in one form or the other.